### PR TITLE
Add Normalized-GPT Architecture

### LINKED ISSUE #38275: Support for Normalized-GPT architecture
### Model description

Normalized Transformer (nGPT) is a novel architecture with representation learning on the hypersphere. In nGPT, all vectors forming the embeddings, MLP, attention matrices and hidden states are unit norm normalized. The input stream of tokens travels on the surface of a hypersphere, with each layer contributing a displacement towards the target output predictions. These displacements are defined by the MLP and attention blocks, whose vector components also reside on the same hypersphere.

Paper: https://arxiv.org/abs/2410.01131

### Open source status

- [x] The model implementation is available
- [x] The model weights are available

Weights will be shared on HF soon.

### PR DESCRIPTION
**Pull request by shan18**
# What does this PR do?

Adds a new architecture called normalized Transformer (nGPT) ([Paper](https://arxiv.org/abs/2410.01131)).

Fixes #38275 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hi @shan18, we generally don't add or review architectures until a significant pre-trained model exists! The paper and architecture do look cool, but in general we wait until there's a checkpoint that's practically useful for the community before we add the code to `transformers`.

**Comment by Rocketknight1** [comment]
@shan18 understood! In that case, you can fix a lot of the failing tests with a few simple fixes:

1) The `setup_and_quality` tests usually require code styling. You can do `pip install -e .[quality]` followed by `make fixup` or `make style` to get those to pass.
2) Several of the failing torch tests look like simple missing imports. You may need to import `require_torch_sdpa` in some test files.

**Comment by shan18** [comment]
@Rocketknight1 the tests have passed. The `tests_processors` shows failed but in the logs I see:
```
================ 234 passed, 219 skipped, 7 warnings in 53.65s =================
```

Not sure how to fix this.

**Comment by shan18** [comment]
Closing this PR for now: we’ve decided to retrain the model from scratch with updated data and some architectural tweaks. We’ll open a new PR once the new checkpoint and code are ready. Thank you for your time and feedback so far.
